### PR TITLE
[10_2_X] Fall17 V2 Photon ID and PhotonIDValueMapProducer speedup

### PR DESCRIPTION
**Pull request by guitargeek**
Primarily, this is the backport of #25293 plus changes to make Photon MVA V2 work before #24131 and adapting to new weights files names.

Additionally, since the NanoAOD developers would really appreciate the PhotonIDValueMapProducer speedup from #25092 to be backported to 10_2_X. Here, I suggest to simply synchronize the PhotonIDValueMapProducer (and required new header files) with master to get the speedup.

Validation plots for all photon ID value maps:
https://rembserj.web.cern.ch/rembserj/plots/CMSSW_PRs/25372/
  * target: CMSSW_10_2_6 plus this PR and the external PR
  * reference: CMSSW_10_4_X_2018-12-03-1100 and CMSSW_10_2_6

Validation workflow:
  * Dump all relevant ID value maps to ROOT file with this configuration [1], running on DY+Jets
  * Compare ROOT file from reference and target with this script [2]

Depends on cms-sw/cmsdist#4518.

[1] https://github.com/guitargeek/cmssw/blob/f76b6a2c9425d5ad5df91df5ef04272d0ab40e39/RecoEgamma/EgammaTools/test/testPhotonIDs_cfg.py
[2] https://github.com/guitargeek/cmssw/blob/f76b6a2c9425d5ad5df91df5ef04272d0ab40e39/RecoEgamma/EgammaTools/test/photon_validation_10_2_6.py

### CODE REVIEW

**Comment by peruzzim** [comment]
@guitargeek I leave to reconstruction and release managers the discussion on how to do the backport technically.

From the NanoAOD production perspective, the part that is on top of #25347 should be completely transparent.
Looking at commits it seems that this PR builds on top of that one at fde7b8ed6376149de42b15161e3e6837a07e1219, does not just depend on it.

Thanks

**Comment by guitargeek** [comment]
Hi Marco, yep it will be discussed today in the RECO meeting. But can you please explain again what you mean with "the part that is on top of #25347 should be completely transparent"? Thanks!

**Comment by guitargeek** [comment]
Hi @slava77 and @perrotta, so did we agree in the end on the minimal backport solution actually? For 10_2_X, that would mean https://github.com/cms-sw/cmssw/pull/25347 and this PR which I'd kindly ask to get tested.

About the minimal backport of V2 IDs to 9_4_X: I was not entirely correct with my assessment this afternoon that it would only require new config files. In reality, the variables are slightly different, as the V2 ID takes the absolute value of track-cluster matching variables [2] while V1 does not [1]. That means an additional `if` here [3].

I suppose I'll just quickly set up the minimal backport to 9_4_X and make a PR, so it can be further discussed there.

Jonas

[1] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/data/ElectronMVAEstimatorRun2Fall17V1Variables.txt#L36
[2] https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/ElectronIdentification/data/ElectronMVAEstimatorRun2Variables.txt#L36
[3] https://github.com/cms-sw/cmssw/blob/CMSSW_9_4_X/RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Fall17.cc#L357

**Comment by guitargeek** [comment]
Ok @perrotta, @slava77, @Sam-Harper and @michelif. I have the minimal backport set up now [1].

In the end, it was not so trivial as expected, since we also need a new cut class for the cut ID and for the electron MVA, I need also to produce the BDT output before the logistic transformation now. Please tell me which option you prefer and I will validate it and make the backport PR:

* Option 1: make minimal changes necessary to get V2 IDs [1].
* Option 2: backport all Egamma ID developments up to the FWlite support [2].

[1] https://github.com/guitargeek/cmssw/tree/EgammaID_9_4_12_minimal
[2] https://github.com/guitargeek/cmssw/tree/EgammaID_FWlite_949

**Comment by perrotta** [comment]
@guitargeek about the backport to 94X, I think we agreed on the "minimal support", and I would concentrate on it by now. The "full backport" can be arranged for a later stage, if needed.

About currently open pull request for 94X: since we agreed about backporting both V2 IDs and the code speed up, this is the pull request which contains them. 

Back to 10_2_X: at this point, #25313 is superseeded, and #25347 is only a subsample of this one, and therefore redundant. I think you can close both them, now.

**Review by perrotta — COMMENTED** [review]
You can find below a few observation from the comparison between what committed in this PR and what was merged in the master. It was not easy to make such a comparison due to the several other changes that were integrated in the same code in the master, and I could have overlooked something: please have a look.

**Comment by guitargeek** [comment]
Thanks for the comments @perrotta, it seems you prefer things to be in sync with master if we change them already. I changed the config files as you suggested and synced the PhotonIDValueMapProducer with master, which required me to also copy the MultiToken and EgammaTools/Utils from master.

**Comment by guitargeek** [comment]
Maybe that should get tested soon? @perrotta, @slava77

**Comment by perrotta** [comment]
please test with cms-sw/cmsdist#4518

**Comment by perrotta** [comment]
+1
- It backports to 10_2_X:
  - the updated MVA and cut based photon IDs from #25293;
  - the PhotonIDValueMapProducer speedup from #25092;
  - additional commits needed to merge the two above in 10_2_X.
- Changes in the miniAOD outputs are similar to what observed when integrated the same photon IDs in #25293 (see https://github.com/cms-sw/cmssw/pull/25293#issuecomment-441808504) and correspond to the additional cut and MVA based IDs
- Other jenkins comparisons show no differences
- It needs cms-sw/cmsdist#4518

**Comment by guitargeek** [comment]
Nice! Thanks Andrea

**Comment by andrius-k** [comment]
+1

**Comment by peruzzim** [comment]
@fabiocos @slava77 it would be urgent to merge this functionality, if possible. Thanks!

**Comment by perrotta** [comment]
- Changes in the miniAOD output are only in the PhotonIDs
- It is expected that only the new IDs are added to the list, while all other remain untouched. Due to the way they are stored in the event, it is hard to judge whether this is actually the case without a dedicated tool that extract any of them separately
- Just to be 100% sure: @guitargeek, could you please provide plots with evidence that none of the other IDs is touched, and only the new ones are added?

**Comment by perrotta** [comment]
hold

**Comment by guitargeek** [comment]
Okay, added. Of course one can't compare the V2 IDs then, because they were not available before. It's in a subdirectory of my validation plots page:

https://rembserj.web.cern.ch/rembserj/plots/CMSSW_PRs/25372/vs_10_2_6/

**Comment by perrotta** [comment]
unhold
(thank you Jonas: by the way, could you please edit the PR description again and write down which workflow you used for your comparison, just for the record?)

**Comment by guitargeek** [comment]
Sure! It's just a EDAnalyzer which dumps all ID value maps and a little python script to make the plots in one of my private CMSSW branches. I reference them in the PR description now.

**Comment by peruzzim** [comment]
@guitargeek this PR needs to be modified to only add the Fall17 V2 photon IDs to miniAOD content (in `PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py`) when one of these modifiers is active:
- run2_miniAOD_devel
- run2_nanoAOD_102Xv1

will depend on #25444

**Comment by perrotta** [comment]
assign xpog

**Comment by peruzzim** [comment]
@guitargeek this is urgent

**Comment by guitargeek** [comment]
Hi @peruzzim, @perrotta, sorry for being off the radar for a while. I committed a a solution to what you requested and merged the branch with the new eras for easier testing.

Although I am very certain this is modifying all the relevant parts of the configuration, I'm not sure if the mechanism I use there will work, as my local tests are still running.

So please, before letting the bot test this PR again, just have a quick look at the python code I added to see if it reasonable.

**Comment by peruzzim** [comment]
@guitargeek IMO you should really avoid introducing further changes, like you have done integrating #25444 in this PR. In this way it becomes very difficult to understand and validate things: as a consequence, we never converge. We have to keep things simple to streamline the integration process, and profit from the tools we have (like testing with another PR).

Then, I don't think it's acceptable to create a temporary cms.Process in a production code.
I would propose you do something like:
```
photon_id_config = cms.PSet(photon_ids = cms.vstring(['id1','id2',...]))
modifier.toModify(photon_id_config, photon_ids = photon_id_config.photon_ids + ['id3',id4'])
```
and then later:
```
for idmod in  photon_id_config.photon_ids.value():
   setupAllVIDIdsInModule...
```

File: PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py [hunk 1]
```diff
@@ -317,11 +317,22 @@ def miniAOD_customizeCommon(process):
 
     #VID Photon IDs
     process.patPhotons.addPhotonID = cms.bool(True)
-    photon_ids = ['RecoEgamma.PhotonIdentification.Identification.cutBasedPhotonID_Fall17_94X_V1_TrueVtx_cff',
+    photon_ids = [
```
Comment (peruzzim) [inline PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py]: please remove this empty list

**Comment by guitargeek** [comment]
Hi Marco, thanks for the advice! I agree that wrapping the list of IDs in a PSet is a good idea, then it can be modified directly.

About merging #25444: well it was just a merge, so after the other PR is merged, this PR simply doesn't show these commits anymore. But the actual reason I did it because we once had problems letting the bot test with two external PRs, so I was unsure if it would work with one external PR and one cmssw PR. Anyway, I took these commits out now and let's hope the testing like that will work.

I'll quickly remove the empty list.

[1] https://github.com/cms-sw/cmssw/pull/25408#issuecomment-444030535

**Comment by peruzzim** [comment]
please test with #25444 cms-sw/cmsdist#4518

**Comment by peruzzim** [comment]
please abort

File: PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py [hunk 2]
```diff
@@ -317,11 +317,20 @@ def miniAOD_customizeCommon(process):
...
                   'RecoEgamma.PhotonIdentification.Identification.cutBasedPhotonID_Spring16_V2p2_cff',
-                  'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring16_nonTrig_V1_cff']
+                  'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring16_nonTrig_V1_cff',
+                     ]))
+
+    from Configuration.Eras.Modifier_run2_miniAOD_devel_cff import run2_miniAOD_devel
+    run2_miniAOD_devel.toModify(photon_id_config, photon_ids = photon_id_config.photon_ids + \
+            ["RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Fall17_94X_V2_cff",
+             "RecoEgamma.PhotonIdentification.Identification.cutBasedPhotonID_Fall17_94X_V2_cff"]
+
```
Comment (peruzzim) [inline PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py]: parenthesis missing?

Comment (guitargeek) [reply PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py]: Yeah I saw that too now after I tried to test locally... thanks.

**Comment by guitargeek** [comment]
Okay @peruzzim , I tested some workflows with MiniAOD locally and they pass. It should be safe now to test with #25444 cms-sw/cmsdist#4518. Sorry for the initial problems with this.

**Comment by peruzzim** [comment]
please test with #25444 cms-sw/cmsdist#4518

**Comment by peruzzim** [comment]
+xpog

for the change to `PhysicsTools/NanoAOD/python/photons_cff.py` that is just a file naming change

**Comment by andrius-k** [comment]
+1

**Comment by perrotta** [comment]
+1
- Changes since last reco signature in https://github.com/cms-sw/cmssw/pull/25372#issuecomment-443896460 are only in ```PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py```: they allow the new IDs to be produced only for the ```run2_miniAOD_devel``` era, and as such there are now no changes at all in the outputs outside that era
- All other features remain unchanged since last reco signature

**Comment by fabiocos** [comment]
merge

**Pull request merged**